rllib/algorithms/algorithm.py

```diff
@@ -941,6 +943,199 @@ def duration_fn(num_units_done):
         # Also return the results here for convenience.
         return self.evaluation_metrics
 
+    @ExperimentalAPI
+    def _evaluate_v2(
+        self,
+        duration_fn: Optional[Callable[[int], int]] = None,
+    ) -> dict:
+        """Evaluates current policy under `evaluation_config` settings.
+
+        Note that this default implementation does not do anything beyond
+        merging evaluation_config with the normal trainer config.
+
+        Args:
+            duration_fn: An optional callable taking the already run
+                num episodes as only arg and returning the number of
+                episodes left to run. It's used to find out whether
+                evaluation should continue.
+        """
+        # How many episodes/timesteps do we need to run?
+        # In "auto" mode (only for parallel eval + training): Run as long
+        # as training lasts.
+        unit = self.config["evaluation_duration_unit"]
+        eval_cfg = self.config["evaluation_config"]
+        rollout_len = eval_cfg["rollout_fragment_length"]
+        num_envs = eval_cfg["num_envs_per_worker"]
+        auto = self.config["evaluation_duration"] == "auto"
+        duration = (
+            self.config["evaluation_duration"]
+            if not auto
+            else (self.config["evaluation_num_workers"] or 1)
+            * (1 if unit == "episodes" else rollout_len)
+        )
+
+        # Call the `_before_evaluate` hook.
+        self._before_evaluate()
+
+        # Put weights only once into object store and use same object
+        # ref to synch to all workers.
+        self._evaluation_weights_seq_number += 1
+        weights_ref = ray.put(self.workers.local_worker().get_weights())
+        self._sync_filters_if_needed(
```

Comment:
can you add an empty line before this? _sync_filters_if_needed() logically is a different piece of work than syncing weights.
also, this may hang for 60 mins actually, if there is at least one hanging worker.
add a TODO for me to sort this out, when we move filters into connectors?

Reply:
@avnishn I think we could introduce another mode of operation for the async request manager for such blocking operations like `sync_filters_if_needed` that checks for the health of actors with a dummy function while calling.

`sync_filters_if_needed` is only called once or twice per training iteration. Such calls are very distinct from our experience sampling calls in that they are very few and small, but all of them should go through if RolloutWorkers are alive, as opposed to our sampling calls which are many and heavy, and we don't care if 1-2 don't make it in time.

So another mode we could introduce here would be calling a function like `_sync_filters_if_needed` and pinging with `actor.apply.remote(lambda_fn)` in parallel for example every minute. If the ping times out, we know at least that the targeted worker is dead and we can disregard the request and move on. 

I know this solution does not sound well engineered, but I think it could be a KISS way of dealing with this because I suppose that most of the time we risk waiting 60 minutes, the worker will simply be dead. It would also add very little logic and would be able to send meaningful warnings.

Reply:
This kind of relates to our current `_worker_health_check`.

Reply:
so the thing to consider here is that our async requests manager will drop requests that go over the max number of inflight requests threshold.